R/signal_noise.R

```diff
@@ -0,0 +1,53 @@
+#' Calculate signal and noise metrics for Open Specy objects
+#'
+#' This function calculates common signal and noise metrics for Open Specy objects.
+#'
+#' @param object An Open Specy object.
+#' @param return A character specifying the desired metric to calculate. Options include "signal" (mean intensity), "noise" (standard deviation of intensity), "signal_times_noise" (absolute value of signal times noise), "signal_over_noise" (absolute value of signal / noise), or "total_signal" (total signal = signal * number of data points).
+#' @param na.rm Logical value indicating whether missing values should be removed when calculating signal and noise. Default is \code{TRUE}.
+#'
+#' @return A numeric vector containing the calculated metric for each spectrum in the Open Specy object.
+#'
+#' @examples
+#' data("raman_hdpe")
+#' signal_noise(raman_hdpe, return = "signal")
+#' signal_noise(raman_hdpe, return = "noise")
+#' signal_noise(raman_hdpe, return = "signal_times_noise")
+#'
+#' @importFrom TTR runMax
+#'
+#' @export
+ signal_noise <- function(object, return = "signal_over_noise", na.rm = TRUE){
+     
+     vapply(object$spectra, function(intensity){
+         if(length(intensity[!is.na(intensity)]) < 20){
+             stop("Need at least 20 intensity values to calculate the signal or noise values accurately.")
+         }
+         if(return == "run_signal_over_noise"){
+         max  = runMax(intensity[!is.na(intensity)], n = 20)
```

Comment:
The TTR package is used only once here for the `runMax` function. Do those calculations need to run super fast? If not, I might code our own runmax function in R instead of using this C++ implementation here to limit the number of package dependencies.

Reply:
They don't need to run extremely fast, they aren't very computationally heavy.